python/src/iceberg/transforms.py

```diff
@@ -0,0 +1,446 @@
+                    (v.int >> 64) & 0xFFFFFFFFFFFFFFFF,
+                    v.int & 0xFFFFFFFFFFFFFFFF,
+                )
+            ),
+            lambda t: t == UUIDType,
+        ),
+        # bucketing by Float/Double is not allowed by the spec, but they have hash implementation
+        FloatType: (
+            lambda v: mmh3.hash(struct.pack("d", v)),
+            lambda t: t == FloatType,
+        ),
+        DoubleType: (
+            lambda v: mmh3.hash(struct.pack("d", v)),
+            lambda t: t == DoubleType,
+        ),
+    }
+
+    def __init__(self, source_type: Type, num_buckets: int):
+        if (
+            source_type not in Bucket._FUNCTIONS_MAP
+            and not isinstance(source_type, FixedType)
+            and not isinstance(source_type, DecimalType)
+        ):
+            raise ValueError(f"Cannot bucket by type: {source_type}")
+
+        super().__init__(
+            f"bucket[{num_buckets}]",
+            f"transforms.bucket(source_type={repr(source_type)}, num_buckets={num_buckets})",
+        )
+        self._type = source_type
+        self._num_buckets = num_buckets
+
+        if isinstance(self._type, FixedType):
+            self._hash_func = lambda v: mmh3.hash(v)
+            self._can_transform = lambda t: isinstance(t, FixedType)
+        elif isinstance(self._type, DecimalType):
+            self._hash_func = lambda v: mmh3.hash(transform_util.decimal_to_bytes(v))
+            self._can_transform = lambda t: isinstance(t, DecimalType)
+        else:
+            self._hash_func = Bucket._FUNCTIONS_MAP[self._type][0]
+            self._can_transform = Bucket._FUNCTIONS_MAP[self._type][1]
+
+    @property
+    def num_buckets(self) -> int:
+        return self._num_buckets
+
+    def apply(self, value) -> Optional[int]:
+        if value is None:
+            return None
+
+        return (self._hash_func(value) & Bucket._MAX_32_BITS_INT) % self._num_buckets
+
+    def can_transform(self, target: Type) -> bool:
+        return self._can_transform(target)
+
+    def result_type(self, source: Type):
+        return IntegerType
+
+
+class Time(Transform):
+    """
+    Time class is for both Date transforms and Timestamp transforms.
+    """
+
+    _TIME_SATISFIED_ORDER = dict(year=3, month=2, day=1, hour=0)
+    _VALID_TIME_GRANULARITY = {
+        DateType: {"year", "month", "day"},
+        TimestampType: {"year", "month", "day", "hour"},
+        TimestamptzType: {"year", "month", "day", "hour"},
+    }
+    _INSTANCES: dict = {DateType: {}, TimestampType: {}, TimestamptzType: {}}
+
+    def __new__(cls, source_type: Type, name: str):
+        if cls._INSTANCES.get(source_type, {}).get(name) is None:
+            if name not in Time._VALID_TIME_GRANULARITY.get(source_type, {}):
+                raise ValueError(f"Cannot partition type: {source_type} by {name}")
+            cls._INSTANCES[source_type][name] = super(Time, cls).__new__(cls)
+        return cls._INSTANCES[source_type][name]
```

Comment:
This seems overly complicated. Why not just have a base class and then fill in what you need to for a few named instances like what is done in `types`?

Reply:
SG, created instances and also pass the time diff functions to it.